tiledb/sm/query/readers/dense_reader.cc

```diff
@@ -468,46 +472,76 @@ Status DenseReader::dense_read() {
         continue;
       }
 
+      std::vector<FilteredData> filtered_data;
       if (condition_names.count(name) == 0) {
         // Read and unfilter tiles.
         to_read[0] = name;
-        RETURN_CANCEL_OR_ERROR(
-            read_and_unfilter_attribute_tiles(to_read, result_tiles));
+        filtered_data = std::move(read_attribute_tiles(to_read, result_tiles));
       }
 
-      // Only copy names that are present in the user buffers.
-      if (buffers_.count(name) != 0) {
-        // Copy attribute data to users buffers.
-        auto& var_buffer_size = var_buffer_sizes[name];
-        status = copy_attribute<DimType, OffType>(
-            name,
-            tile_extents,
-            subarray,
-            t_start,
-            t_end,
-            subarray_start_cell,
-            subarray_end_cell,
-            tile_subarrays,
-            tile_offsets,
-            var_buffer_size,
-            range_info,
-            result_space_tiles,
-            qc_result,
-            num_range_threads);
-        RETURN_CANCEL_OR_ERROR(status);
+      if (compute_task.has_value()) {
+        compute_task->wait();
+        RETURN_NOT_OK(compute_task->get());
+        compute_task = nullopt;
+
+        if (read_state_.overflowed_) {
+          return Status::Ok();
+        }
       }
 
-      clear_tiles(name, result_tiles);
-    }
+      compute_task = storage_manager_->compute_tp()->execute(
```

Comment:
Yes, it might not be used at all, in which case we don't have to wait for it to complete anywhere.

Reply:
Yes, for the first iteration, there is no task so we don't need to wait for it.

Reply:
You can call `valid()` on a future to see if a value has been set on it.  I think the best reason to use an `optional` (though `shared_ptr` would also work) is that you can do ```compute_task = ...``` and get a new future in `compute_task`.  You can't do ```compute_task =...``` if `compute_task` *is* a future. 

You do have to be a little careful in doing ```compute_task =...``` if `compute_task` is wrapping a future that you haven't called `wait` or `get` on.  There are some nitpicks rules about what happens to a future when its destructor is called -- it can block in fact -- which gets us back into the danger zone for deadlock.  However, if you are careful to always `wait` or `get` the future in `compute_task` before filling it again, we will be okay.  Though it makes me kind of nervous.  Maybe have an explicit flag to indicate first time through (express intent) and clear `compute_task` after every `wait` or `get`.  Then you can check that it is empty every place before you assign to it.